src/Features/Core/Portable/EditAndContinue/EditAndContinueMethodDebugInfoReader.cs

```diff
@@ -0,0 +1,157 @@
+                    else
+                    {
+                        localSlots = lambdaMap = default;
+                    }
+
+                    return EditAndContinueMethodDebugInformation.Create(localSlots, lambdaMap);
+                }
+                catch (InvalidOperationException e) when (FatalError.ReportWithoutCrash(e)) // likely a bug in the compiler/debugger
+                {
+                    // TODO: CustomDebugInfoReader should throw InvalidDataException
+                    throw new InvalidDataException(e.Message, e);
+                }
+            }
+        }
+
+        private sealed class Portable : EditAndContinueMethodDebugInfoReader
+        {
+            private readonly MetadataReader _pdbReader;
+
+            public Portable(MetadataReader pdbReader)
+            {
+                _pdbReader = pdbReader;
+            }
+
+            public override bool IsPortable => true;
+
+            public override StandaloneSignatureHandle GetLocalSignature(MethodDefinitionHandle methodHandle)
+                => _pdbReader.GetMethodDebugInformation(methodHandle.ToDebugInformationHandle()).LocalSignature;
+
+            public override EditAndContinueMethodDebugInformation GetDebugInfo(MethodDefinitionHandle methodHandle)
+                => EditAndContinueMethodDebugInformation.Create(
+                    compressedSlotMap: GetCdiBytes(methodHandle, PortableCustomDebugInfoKinds.EncLocalSlotMap),
+                    compressedLambdaMap: GetCdiBytes(methodHandle, PortableCustomDebugInfoKinds.EncLambdaAndClosureMap));
+
+            private ImmutableArray<byte> GetCdiBytes(MethodDefinitionHandle methodHandle, Guid kind)
+                => TryGetCustomDebugInformation(_pdbReader, methodHandle, kind, out var cdi) ?
+                    _pdbReader.GetBlobContent(cdi.Value) : default;
+
+            /// <exception cref="BadImageFormatException">Invalid data format.</exception>
+            private static bool TryGetCustomDebugInformation(MetadataReader reader, EntityHandle handle, Guid kind, out CustomDebugInformation customDebugInfo)
+            {
+                bool foundAny = false;
+                customDebugInfo = default;
+                foreach (var infoHandle in reader.GetCustomDebugInformation(handle))
+                {
+                    var info = reader.GetCustomDebugInformation(infoHandle);
+                    var id = reader.GetGuid(info.Kind);
+                    if (id == kind)
+                    {
+                        if (foundAny)
+                        {
+                            throw new BadImageFormatException();
+                        }
+
+                        customDebugInfo = info;
+                        foundAny = true;
+                    }
+                }
+
+                return foundAny;
+            }
+        }
+
+        public unsafe static EditAndContinueMethodDebugInfoReader Create(ISymUnmanagedReader5 symReader, int version)
```

Comment:
>public unsafe static EditAndContinueMethodDebugInfoReader Create(ISymUnmanagedReader5 symReader, int version) [](start = 8, length = 109)

nit: better move the public method above private classes

Reply:
We usually have classes first.